app/src/main/java/com/firebase/uidemo/database/firestore/FirestorePagingActivity.java

```diff
@@ -0,0 +1,194 @@
+                .build();
+
+        FirestorePagingAdapter<Item, ItemViewHolder> adapter =
+                new FirestorePagingAdapter<Item, ItemViewHolder>(options) {
+                    @NonNull
+                    @Override
+                    public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
+                                                             int viewType) {
+                        View view = LayoutInflater.from(parent.getContext())
+                                .inflate(R.layout.item_item, parent, false);
+                        return new ItemViewHolder(view);
+                    }
+
+                    @Override
+                    protected void onBindViewHolder(@NonNull ItemViewHolder holder,
+                                                    int position,
+                                                    Item model) {
+                        holder.bind(model);
+                    }
+
+                    @Override
+                    protected void onLoadingStateChanged(@NonNull LoadingState state) {
+                        switch (state) {
+                            case LOADING_INITIAL:
+                            case LOADING_MORE:
+                                mProgressBar.setVisibility(View.VISIBLE);
+                                break;
+                            case LOADED:
+                                mProgressBar.setVisibility(View.GONE);
+                                break;
+                            case ERROR:
+                                showToast("An error occurred.");
+                                break;
+                        }
+                    }
+                };
+
+        mRecycler.setLayoutManager(new LinearLayoutManager(this));
+        mRecycler.setAdapter(adapter);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.menu_firestore_paging, menu);
+        return super.onCreateOptionsMenu(menu);
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        if (item.getItemId() == R.id.item_add_data) {
+            showToast("Adding data...");
+            createItems().addOnCompleteListener(this, new OnCompleteListener<Void>() {
+                @Override
+                public void onComplete(@NonNull Task<Void> task) {
+                    if (task.isSuccessful()) {
+                        showToast("Data added.");
+                    } else {
+                        Log.w(TAG, "addData", task.getException());
+                        showToast("Error adding data.");
+                    }
+                }
+            });
+
+            return true;
+        }
+        return super.onOptionsItemSelected(item);
+    }
+
+    private Task<Void> createItems() {
+
+        WriteBatch writeBatch = mFirestore.batch();
+        CollectionReference collRef = mFirestore.collection("items");
+
+        for (int i = 0; i < 500; i++) {
```

Comment:
Isn't 500 a bit excessive for a page size of 20? Maybe like 250 or something?

Also, while I've snagged your attention, the pricing for write batches really isn't clear in the [docs](https://firebase.google.com/docs/firestore/pricing) (no mention, actually). You're still paying for 500 additions, right?

Reply:
Re: pricing yeah they are priced per-write.

Reply:
Re: re: pricing thanks! 😁